src/components/button/button.tsx

```diff
@@ -54,6 +55,14 @@ export class Button {
     @Prop({ reflect: true })
     public loading = false;
 
+    /**
+     * Set to `true` to make the button automatically hide when disabled.
+     * While loading or showing the checkmark after loading, the button will
+     * still be visible.
+     */
+    @Prop({ reflect: true })
+    public hideWhenDisabled = false;
```

Comment:
My main argument against this is that it makes the API less clean. If the consumer wants to hide the button they can just hide it themselves. They know when they want to disable it and therefore also know when they do not want to display it. 

Also, why even set it to `disabled` in the first place if they are going to hide it? :thinking: `disable` will loose it's meaning since the button is not visible anyway. Another approach would be to use the global html `hidden` attribute instead if they want to hide it. Thoughts? :slightly_smiling_face:

Reply:
The main reason was to not have the consumer account for the time the "loading done" checkmark is shown after loading. Granted, we should give the consumer some more control over that, because it really should only be shown if something was being saved or sent or similar, and the process was successful.

But, handing off the *whole* thing to the consumer won't really work, because the checkmark isn't just a different icon for the label, it's replacing the label, and there are fading transitions between the label, spinner, checkmark, and back to the label again. And we want the timing to consistent in every place the checkmark is shown.

We can take care of all of those things, and present them to the consumer through a very simple interface, where they essentially just "tick the boxes" for the features they want, and don't have to care about timing and transitions and stuff.

Reply:
I agree with @adrianschmidt. The benefits are that:
1. We're in control, if anything changes in future. Let's say an animation length changes. Or the progress reflects the actual time the process is taking.
1. Consumers don't have to think and find solutions. They just say I want this feature. All they have to focus on is when to make the button disable.

Reply:
Regarding the HTML attribute `<button hidden>`. It's up to you Adrian. This won't be a CSS-only-fix anymore though, as a script should add the `hidden` attribute, after the animation has run.

Reply:
We can use the `hidden` attribute to hide the button, instead of adding a class `.hidden`.

But we can't use the `hidden` attribute instead of the `hideWhenDisabled` because they do different things. `hideWhenDisabled` doesn't hide the button. It just tells the button that when it's disabled, *if* there is no other information it needs to convey, like showing the spinner or the checkmark, then it should hide itself.

Reply:
Personally I don't think it's a problem that the API becomes larger, since these more complicated options are optional. But we could provide a `limel-button` with a much simpler API, and a `limel-complex-button` or something, that provides all the special options…

Reply:
Hmm ok, I had forgotten about all the spinner stuff I think, and I'm not even sure if I get how all that works :smile: 

I agree that it's too much work for the consumer to take care of that, they should not have to worry about the timing of things etc. I'm also not against making the API larger, I'm mostly against mixing concepts that seem unrelated :slightly_smiling_face: Maybe it's just a naming thing? Maybe we can find a better name for the property, because I still think that `disabled` and `hidden` are two separate things and should not be coupled like this :thinking: Why set the state to disabled when you want to hide it?

Reply:
Ah, now I understand where you're coming from 😄

If your primary goal is to hide the button, then no, you shouldn't use this property. Then you should just set the "native" `hidden` attribute.

Instead, think of this as setting an alternative disabled-style for the button. Don't think of this as setting a *state*, think of it as setting a *type* of button, just like `primary`. And `disabled` shouldn't be set to hide the button, `disabled` should be set to disable the button, and, if appropriate based on the circumstances, the alternative disabled-styling will be applied.

Reply:
We can definitely rename the property, if we can think of a better name 👍

Reply:
I had another thought about this yesterday. Do we never want to show a disabled button on the card? What about when the form is invalid? We are not using the validation status of the form at the moment, but if we do that do we want to hide the button when the form has been edited but is invalid or do we want to show the disabled button instead? Or should we display some other indicator that the form has been altered but is invalid? I guess this might affect how we implement this :thinking:

Reply:
Yeah, you are absolutely correct. If the form is invalid, we want to disable the button, but keep it visible. So if we would go with my implementation, there would be cases where the consumer would flip `hideWhenDisabled` to `false`.

I understand your concerns, and I think I mostly agree, but I can't come up with a better alternative at the moment. Do you have any ideas?

Reply:
I think it's good if we keep things as simple as possible and let the consumer be in control. We cannot stop anyone from using the `hidden` attribute and that will hide the button no matter what (I think?), so if we want to have a nicer experience and be in control in regards to animations, spinners etc. we have to use something else. What about just having something like `show` or `hide`? Setting that could hide the button but only when all the animations etc. are done. Will that solve our problems? :slightly_smiling_face:

Reply:
So, basically just cut `WhenDisabled` from `hideWhenDisabled`? I'm not sure that's any less confusing, but yeah, I guess…

Maybe we should call it something completely different, and maybe not even add it as a property, but as a helper-class. Similar to `has-low-density` and `has-interactive-rows` on limel-table?

Something like `has-reduced-presence`?

Reply:
Yeah that might be a better idea :slightly_smiling_face:

Reply:
or something like `hides-when-not-needed` 🙃

Reply:
I would have liked `hides-when-not-needed` if the button itself would always know if it's needed or not. But it won't, unless we give it knowledge of the form and its validity. I'd like to avoid that 🙂

Reply:
I've renamed the property, but haven't changed it to a helper-class yet.
I'll add another fixup for that, and we can see which one we like better.

Reply:
Ah, no, making it into a helper-class makes things more difficult, because we can't just style directly off of that, because of the shadowDom boundary… so we'd have to set up some javascript to look for that class, and since the class may have to be removed and added again at times, we'd have to have some kind of watcher and stuff.

Unless I'm forgetting something because it's Friday afternoon, I think it's best we just go with the property.

Reply:
Hmm is that true? :thinking: 
I don't think we added any code for the classes that we added for the list, I think this should work even if we have the shadow boundary

Reply:
I'm not sure… I'll have a look at the tables 😄

Reply:
What the heck?!

![image](https://user-images.githubusercontent.com/154725/99083856-6bf67b80-25c6-11eb-9174-77904fe3b3c0.png)

I didn't think it was possible to nest stuff inside the `:host` selector…

I swear I've tried using

```scss
:host {
	display: box;

	&[hidden] {
		display: none;
	}
}
```

But maybe it's just the `&` that doesn't work for the `:host`… 🤔

I've been living a LIE!!! 🤦‍♂️

Reply:
Makes me the only problem was that I didn't add parenthesis… since I didn't do that in the example above… 😭

Reply:
⚡

Reply:
Yeah I think  you need those parentheses for some reason :upside_down_face: